src/optlang/matrix_interface.py

```diff
@@ -0,0 +1,961 @@
+
+1. Create a class that inherits from MatrixProblem and implement the `solve` method
+   and optionally overwrite the `reset` and `still_valid` methods.
+
+2. Create derived classes for Variable, Constraint, Objective, and Configuration.
+   You don't need to overwrite anything. Those should work out of the box.
+
+3. Create a derived class for Model and set the `ProblemClass` attribute to the class
+   from (1) and the `status_map` to map the solver status to an Optlang status from
+   `optlang.interface`.
+
+For an example take a look at `optlang.osqp_interface`.
+"""
+
+import logging
+import six
+import pickle
+from collections import defaultdict
+from numpy import array, concatenate, Infinity, zeros
+
+from optlang import interface, symbolics, available_solvers
+from optlang.util import inheritdocstring
+from optlang.expression_parsing import parse_optimization_expression
+from optlang.exceptions import SolverError
+
+from scipy.sparse import csc_matrix
+
+from optlang.symbolics import add, mul
+
+log = logging.getLogger(__name__)
+
+
+_STATUS_MAP = defaultdict(lambda: interface.UNDEFINED)
+
+_LP_METHODS = ("auto", )
+
+_QP_METHODS = ("auto", )
+
+_TYPES = ("continuous", "binary", "integer")
+
+
+class MatrixProblem(object):
+    """A concise representation of an LP/QP/MILP problem.
+
+    This class assumes that the problem will be pretty much immutable. This is
+    a small intermediate layer based on dictionaries that is fast to modify
+    but can also be converted to a matrix formulation quickly.
+    """
+
+    def __init__(self):
+        self.variables = set()
+        self.integer_vars = set()
+        self.constraints = set()
+        self.constraint_coefs = dict()
+        self.constraint_lbs = dict()
+        self.constraint_ubs = dict()
+        self.variable_lbs = dict()
+        self.variable_ubs = dict()
+        self.obj_linear_coefs = dict()
+        self.obj_quadratic_coefs = dict()
+        self.primals = {}
+        self.cprimals = {}
+        self.duals = {}
+        self.vduals = {}
+        self.obj_value = None
+        self.direction = -1
+        self.info = None
+        self.status = None
+        self._solution = None
+        self.default_settings()
+
+    def default_settings(self):
+        """Set the default settings of the solver.
+
+        Returns
+        -------
+        dict
+            The default settings for the solver.
+        """
+        self.settings = {
+            "presolve": True,
+            "lp_method": "auto",
+            "qp_method": "auto",
+            "time_limit": 0,
+            "primal_inf_tolerance": 1e-6,
+            "dual_inf_tolerance": 1e-6,
+            "optimality_tolerance": 1e-6,
+            "mip_tolerance": 1e-6,
+            "verbose": 1,
+            "threads": 1,
+            "iteration_limit": 100000
+        }
+
+    def build(self, add_variable_constraints=False):
+        """Build the problem instance.
```

Comment:
Is it the right approach to always build a matrix like this? Should this not be an abstract method, too, that is implemented by the concrete interface class?

Reply:
Most matrix solvers will want the problem in standard form, so the matrices will always be the same. They might use different formats and not always CSC but it is pretty common. For isntance, that implementation would work for OSQP and cvxpy and also the new hybrid interface (if deemed in scope).

Reply:
Okay, great. I suppose, if a particular solver has special needs, that interface can still override this method anyway.